### PR TITLE
[Tracer][ONNX] Enable numel tracing

### LINKED ISSUE #74047: tensor.numel() traced as constant
### 🐛 Describe the bug

Example for repro:
```python
import torch

class Model(torch.nn.Module):
    def forward(self, x):
        return torch.ones(x.numel())

x = torch.randn(2, 3, 4)
g = torch.jit.trace(Model(), (x, ))
print(g.graph)
```
Log:
```
graph(%self : __torch__.Model,
      %x : Float(2, 3, 4, strides=[12, 4, 1], requires_grad=0, device=cpu)):
  %4 : int = prim::Constant[value=24]() # repro_march_numel.py:5:0
  %5 : int[] = prim::ListConstruct(%4)
  %6 : int = prim::Constant[value=6]() # repro_march_numel.py:5:0
  %7 : NoneType = prim::Constant()
  %8 : Device = prim::Constant[value="cpu"]() # repro_march_numel.py:5:0
  %9 : bool = prim::Constant[value=0]() # repro_march_numel.py:5:0
  %10 : Float(24, strides=[1], requires_grad=0, device=cpu) = aten::ones(%5, %6, %7, %8, %9) # repro_march_numel.py:5:0
  return (%10)
```
Observe that `x.numel()` got traced as `%4 : int = prim::Constant[value=24]()` which has fixed value. A more accurate graph is to dynamically compute from shape of `x`, as how `tensor.size()` is traced.

### Versions

PyTorch version: 1.12.0a0+git1908433
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04) 9.4.0
Clang version: 10.0.0-4ubuntu1
CMake version: version 3.22.2
Libc version: glibc-2.17

Python version: 3.7.11 (default, Jul 27 2021, 14:32:16)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.13.0-28-generic-x86_64-with-debian-bullseye-sid
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1080
Nvidia driver version: 510.47.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.3.2
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] torch==1.12.0a0+git187f733

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74082
* __->__ #74081

Fixes #74047

Trace `tensor.numel()` as `aten::numel`, instead of a constant value. The logic is similar to tracing `tensor.size(dim)`. 
Replaces #74059

Differential Revision: [D35299848](https://our.internmc.facebook.com/intern/diff/D35299848)

### CODE REVIEW

File: test/onnx/expect/TestOperators.test_view_flatten.expect [hunk 1]
```diff
@@ -39,16 +121,16 @@ graph {
...
       tensor_type {
         elem_type: 1
         shape {
           dim {
-            dim_value: 1
+            dim_param: "Reshape11_dim_0"
           }
           dim {
-            dim_value: 24
+            dim_param: "Reshape11_dim_1"
```
Comment (BowenBao) [inline test/onnx/expect/TestOperators.test_view_flatten.expect:133]: Note this dynamic change is introduced by `tensor.numel` being traced dynamically, and will be reverted in the next PR in the stack where constant folding for `onnx::ReduceProd` is introduced.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -6474,6 +6474,9 @@ def test_embedding_bag(self):
         offset = torch.tensor([0, 2, 5, 6])
         self.run_test(model, (input, offset))
 
+    @disableScriptTest()  # error in propagate as assign input shape
+    @skipIfUnsupportedMinOpsetVersion(11)
+    def test_embedding_bag_with_offset(self):
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:6479]: Similar to above, offset becomes dynamic, hence requires opset 11.
Next PR in the stack reverts this effect.

**Comment by BowenBao** [comment]
@albanD please let me know if this looks good to you.

**Comment by albanD** [comment]
@BowenBao there are indeed precedents of special casing for tracing in these functions. So I guess this is ok to do.

**Comment by BowenBao** [comment]
Thanks @albanD .

@malfet please import. Only the first PR needs importing as it modifies tracer, the second one can be auto merged.

**Comment by BowenBao** [comment]
@albanD could you approve on behalf of the JIT changes?

@shubhambhokare1, @garymm please review ONNX side changes, thanks.

File: test/jit/test_tracer.py [hunk 3]
```diff
@@ -377,6 +377,18 @@ def test_trace_size(self):
...
         self.do_trace_size(True)
 
+    def test_trace_numel(self):
+        def fn(x):
+            return x.numel()
+
+        x = torch.randn(2, 3, 4)
+        y = torch.randn(4, 5, 6)
+
+        # Check that it behaves as expected
```
Comment (garymm) [inline test/jit/test_tracer.py:387]: Nit: this comment is pretty useless. All tests do that :-)

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 4]
```diff
@@ -6593,24 +6601,28 @@ def forward(self, input, batch1, batch2, alpha, beta):
...
             def forward(self, input):
                 return input.numel() * input
 
         x = torch.randn(2, 3, 5)
+        x2 = torch.randn(4, 5, 6)
         model = MyModule()
-        self.run_test(model, (x,))
+        self.run_test(model, (x,),
+                      input_names=['x'], dynamic_axes={'x': [0, 1, 2]},
+                      test_with_inputs=[(x2, )])
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:6613]:
```suggestion
                      test_with_inputs=[(x2,)])
```

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 5]
```diff
@@ -6593,24 +6601,28 @@ def forward(self, input, batch1, batch2, alpha, beta):
...
             def forward(self, input):
                 return input.numel() * input
 
         x = torch.randn(0)
+        x2 = torch.randn(4)
         model = MyModule()
-        self.run_test(model, (x,))
+        self.run_test(model, (x,),
+                      input_names=['x'], dynamic_axes={'x': [0]},
+                      test_with_inputs=[(x2, )])
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:6625]:
```suggestion
                      test_with_inputs=[(x2,)])
```

**Review by garymm — APPROVED** [review]
🦈

**Comment by albanD** [comment]
I'm afraid I don't know the tracer code at all, @eellison can you take a quick look to make sure it is ok please?

File: torch/csrc/jit/frontend/tracer.cpp [hunk 6]
```diff
@@ -894,6 +894,27 @@ autograd::Variable getSizeOf(const autograd::Variable& var, int64_t dim) {
   return size_var;
 }
 
+autograd::Variable getNumelOf(const autograd::Variable& var) {
```
Comment (eellison) [inline torch/csrc/jit/frontend/tracer.cpp:897]: Could we refactor this with `getSizeOf` maybe ?

**Review by eellison — APPROVED** [review]
LGTM

File: torch/csrc/jit/frontend/tracer.cpp (hunk 6 above)
Comment (BowenBao) [reply torch/csrc/jit/frontend/tracer.cpp:897]: I feel it would add more complexity for handling different arguments based on operator types, while currently we only need inserting a constant in graph for `dim` of `Size`. 
If we choose to address arguments outside of the function, then we'd need the same boilerplate for getting graph from tracing state. 
Please let me know if you have a better idea.

Comment (eellison) [reply torch/csrc/jit/frontend/tracer.cpp:897]: yea it's fine as is

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35299848).

**Comment by BowenBao** [comment]
@malfet rebased

**Comment by malfet** [comment]
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35299848).
